Lib/test/test_when.py

```diff
@@ -0,0 +1,117 @@
+import unittest
+from when import *
+
+# Initial test cases for when and cowns
+# Note: no concurrency test yet
+class WhenTest(unittest.TestCase):
+    obj = None
+
+    def setUp(self):
+        makeimmutable(self.obj)
+
+    def test_cown(self):
+        def invalid_assignment1(c):
+            c.value = 42
+        def invalid_assignment2(c):
+            c.f = 42
+        def invalid_assignment3(c):
+            c["g"] = 42
+
+        c = Cown()
+        self.assertRaises(AttributeError, invalid_assignment1, c)
+        self.assertRaises(AttributeError, invalid_assignment2, c)
+        self.assertRaises(TypeError, invalid_assignment3, c)
+        # Cannot access unacquired cown
+        self.assertRaises(RegionError, lambda _ : c.get(), c)
+        self.assertRaises(RegionError, lambda _ : c.set(Region()), c)
+
+    def test_cown_aquired_access(self):
+        c = Cown()
+        @when(c)
+        def _(c):
+            c.set(self.obj)
+        @when(c)
+        def _(c):
+            self.assertEqual(c.get(), self.obj)
+
+    def test_when_positional(self):
+        a = 42
+        makeimmutable(a)
+        b = "42"
+        makeimmutable(b)
+
+        @when(a, b)
+        def _(c, d):
+            self.assertEqual(a, c)
+            self.assertEqual(b, d)
+
+    def test_when_positional_and_keywords(self):
+        a = 42
+        makeimmutable(a)
+        b = "42"
+        makeimmutable(b)
+
+        @when(a, kw=b)
+        def _(c, kw=b):
+            self.assertEqual(a, c)
+            self.assertEqual(b, kw)
+
+    # Returns the state of a cown as a string
+    # Hacky but want to avoid adding methods to cowns just for testing
+    def hacky_state_check(self, cown, expected_state):
+        s = repr(cown)
+        return expected_state in s
+
+    def test_release(self):
+        r = Region()
+        c = Cown(r)
+        self.assertFalse(r.is_open())
+        self.assertTrue(self.hacky_state_check(c, "released"))
+
+    def test_early_release(self):
+        c = Cown()
+        @when(c)
+        def _(c):
```

Comment:
Do we need to pass `c`, can't we just capture it in the closure?
```suggestion
        def _():
```
Which I guess leads onto what happens about closures in general.  With synchronous this is easier as it still still the same local region.  Async when would be harder?